### PR TITLE
Rule for zlib inflate/deflate

### LINKED ISSUE #497: rule idea: zlib inflate/deflate
inflate: https://github.com/madler/zlib/blob/cacf7f1d4e3d44d871b605da3b647f07d718623f/inflate.c#L622
deflate: https://github.com/madler/zlib/blob/cacf7f1d4e3d44d871b605da3b647f07d718623f/deflate.c#L763

### PR DESCRIPTION
**Pull request by bkojusner**
This rule passes the thorough linting test and closes #497

### CODE REVIEW

File: data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        - bytes: 10 05 01 00 17 05 01 01 13 05 11 00 1B 05 01 10 = Distance of fixed Huffman table
+        - bytes: 60 07 00 00 00 08 50 00 00 08 10 00 14 08 73 00 = Length of fixed Huffman table
+      - and:
+        - description: Captures artifacts in ZLIB deflate
+        - operand[1].number: 0x2A = INIT_STATE
+        - basic block:
+          - and:
+            - description: Write ZLIB Header
+            - mnemonic: shl
+            - operand[1].number: 0xC = Shift s->w_bits
```
Comment (mr-tz) [inline data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml:28]: for `shl  X, 0xC` you can use the `instruction` scope, same for `sub` below

File: data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        - operand[1].number: 0x49 = NAME_STATE
+        - characteristic: loop
+        - basic block:
+          - and:
+            - description: flush_pending - flushes as much pending deflate output as possible
+            - count(mnemonic(mov)): 4
+            - mnemonic: call
+            - count(mnemonic(add)): 3
+            - count(mnemonic(sub)): 2
+            - mnemonic: jnz
```
Comment (mr-tz) [inline data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml:43]: These seem very specific. Can we:
- use count: `X or more/less`?
- rethink the conditional jump (could the compiler use `jz` or similar?)?

**Review by mr-tz — COMMENTED** [review]
Great, thanks for tackling this! Looks good on a first glance although some parts seem very specific to the test samples. Let's discuss inline.

File: data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml (hunk 1 above)
Comment (bkojusner) [reply data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml]: Thank you for pointing that out, implemented instruction scope on both.

File: data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml (hunk 2 above)
Comment (bkojusner) [reply data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml]: This basic block ended up being too specific so I decided to remove it after some further testing. I can add it back with a  more limited count number if requested. Thanks!

**Review by mr-tz — APPROVED** [review]
LGTM, thanks!

File: data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+              - mnemonic: shl
+              - operand[1].number: 0xC = Shift s->w_bits
+            - instruction:
+              - mnemonic: sub
+              - operand[1].number: 0x7800 = Offset to header
+        - operand[1].number: 0x08 = Z_DEFLATED
+        - operand[1].number: 0x06 = level_flags
+        - operand[1].number: 0x20 = PRESET_DICT
+        - operand[1].number: 0x49 = NAME_STATE
+        - characteristic: loop
```
Comment (williballenthin) [inline data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml:37]: im concerned about FPs here, given that many of these numbers are less than 0x100 and therefore are reasonable offsets within structures (except maybe 0x49 and the other odd/unaligned numbers). makes me wish we had that testbest to be able to show the prevalence of rule/branch matches.

Comment (mr-tz) [reply data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml:37]: I'll run this against a bunch of frozen features.

Comment (mr-tz) [reply data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml:37]: Found 5 matches across 85 samples from our testbed.

```
tests/data/112f9f0e8d349858a80dd8c14190e620.exe_
tests/data/276f691a3df25481f59d79781799e35f.exe_
tests/data/39c05b15e9834ac93f206bc114d0a00c357c888db567ba8f5345da0529cbed41.dll_
tests/data/42e81cc1145ba3c1936a6cf9b8da0ccd.dll_
tests/data/50d5ee1ce2ca5e30c6b1019ee64eeec2.exe_
```

@bkojusner if you have a moment, please review these. I can also take a look if you're busy.

Comment (mr-tz) [reply data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml:37]: All of these were true positives.

Comment (williballenthin) [reply data-manipulation/compression/compress-data-via-zlib-inflate-or-deflate.yml:37]: ok, great! that gives me more confidence that we should merge as-is. thanks!

**Review by williballenthin — APPROVED** [review]

**Pull request merged**